File: scripts/data_pipeline/research_files.py
"""Load relocated research evidence without rewriting frozen files on disk."""
import json
from pathlib import Path

MIGRATION = json.loads(
    (Path(__file__).resolve().parents[2] / "docs/research/path-migration.json").read_text(encoding="utf-8")
)


def parse_research_json(text):
    def relocate(value):
        if isinstance(value, str):
            return MIGRATION["paths"].get(value, MIGRATION["sourceRevisions"].get(value, value))
        if isinstance(value, list):
            return [relocate(entry) for entry in value]
        if isinstance(value, dict):
            return {MIGRATION["paths"].get(key, key): relocate(entry) for key, entry in value.items()}
        return value

    return relocate(json.loads(text))

File: src/engine/research-files.ts
import { readFileSync } from 'node:fs';

const migration: { paths: Record<string, string>; sourceRevisions: Record<string, string> } = JSON.parse(
  readFileSync(new URL('../../docs/research/path-migration.json', import.meta.url), 'utf8'));

function relocate(value: unknown): unknown {
  if (typeof value === 'string') return migration.paths[value] ?? migration.sourceRevisions[value] ?? value;
  if (Array.isArray(value)) return value.map(relocate);
  if (value !== null && typeof value === 'object') {
    return Object.fromEntries(Object.entries(value).map(([key, entry]) => [migration.paths[key] ?? key, relocate(entry)]));
  }
  return value;
}

export function parseResearchJson(text: string): ReturnType<typeof JSON.parse> {
  return relocate(JSON.parse(text));
}